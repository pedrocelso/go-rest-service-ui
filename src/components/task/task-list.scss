.card {
  margin: 0 !important;
  &__grid {
    padding: 0.5rem;
    &-center {
      align-self: center
    }
    
    &__pill {
      margin: auto;
      display: block !important;
    }
  }
}

.incident {
  padding: 0 !important;
  &__paper {
      width: stretch;
      min-width: fit-content;
    }
}