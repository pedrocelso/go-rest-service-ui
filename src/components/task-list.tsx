import React, { Component } from 'react';
import {isEmpty, isNil, map} from 'ramda'

import { TaskService, Task } from '../services/task';
import './task-list.scss'

interface TaskProps {
  service: TaskService;
}

interface TaskState {
  taskList: Task[];
}

class TaskList extends Component<TaskProps, TaskState> {
  constructor(props: TaskProps) {
    super(props)
    this.state = {taskList: []}
  }

  componentWillMount() {
    const {service} = this.props
    service.getTasks()
      .fork(
        () => console.error,
        (taskList) => {
          this.setState({taskList})
        }
      )
  }

  render() {
    const {state} = this
    const {taskList} = state

    const listEl = !isNil(taskList) && !isEmpty(taskList) ? (
      <div className="list-group">
        {
          map(({id, name, description}) => {
            return (<a
              href="#"
              className="list-group-item list-group-item-action"
              key={id}
            >
              {name} - {description}
            </a>)
          }, taskList)
        }
      </div>
    ) : null

    return (
      <div className="panel panel-default">
        {listEl}
      </div>
    )
  }
}

export default TaskList;
