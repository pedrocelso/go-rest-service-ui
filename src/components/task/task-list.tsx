import React, { Component } from 'react';
import { connect } from 'react-redux'
import {isEmpty, isNil, map} from 'ramda'

import { deauthenticate } from '../login/login-actions'
import { TaskService, Task } from '../../services/task';
import { TaskItem } from './task-item'
import './task-list.scss'

interface TaskProps {
  deauthenticate: typeof deauthenticate
  service: TaskService;
}

interface TaskState {
  taskList: Task[];
}

export class TaskList extends Component<TaskProps, TaskState> {
  constructor(props: TaskProps) {
    super(props)
    this.state = {taskList: []}
  }

  componentWillMount() {
    const {service} = this.props
    service.getTasks()
      .fork(
        (e) => {
          if (e.statusCode === 401) {
            this.props.deauthenticate();
            sessionStorage.removeItem(`jwtToken`)
          }
        },
        (taskList) => {
          this.setState({taskList})
        }
      )
  }

  render() {
    const {state} = this
    const {taskList} = state

    const listEl = !isNil(taskList) && !isEmpty(taskList) ? (
      <div className="list-group">
        <div className="list-group-item d-flex row">
          <div className="col-md-7"><h5>Task Details</h5></div>
          <div className="col-md-2"><h5>Created at</h5></div>
          <div className="col-md-2"><h5>Updated at</h5></div>
          <div className="col-md-1"><h5>Inccidents</h5></div>
        </div>
        {map((t) => (<TaskItem task={t} key={t.id}/>), taskList)}
      </div>
    ) : null

    return (
      <div className="panel panel-default">
        {listEl}
      </div>
    )
  }
}

const mapDispatchToProps = { deauthenticate }
export default connect(null, mapDispatchToProps)(TaskList)