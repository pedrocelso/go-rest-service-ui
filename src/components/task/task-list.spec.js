import React from 'react';
import { shallow } from 'enzyme'
import sinon from 'sinon'

import { TaskList } from './task-list'

describe(`<TaskList />`, () => {
  it(`Should fetch the tasks from service and render`, () => {
    const service = {
      getTasks: (_) => ({
        fork: (rej, res) => {
          res([
            { id: 10, name: `test`, description: `test@test.com` },
            { id: 11, name: `test1`, description: `test1@test.com` },
          ])
        }
      })
    }

    const component = shallow(
      <TaskList service={service} deauthenticate={sinon.fake()} />,
    );

    expect(component).toMatchSnapshot();
  });

  it(`Should not display tasks if service returns an error`, () => {
    const service = {
      getTasks: (_) => ({
        fork: (rej, _) => {
          rej(`error fetching tasks`)
        }
      })
    }

    const component = shallow(
      <TaskList service={service} deauthenticate={sinon.fake()} />,
    );

    expect(component).toMatchSnapshot();
  });

  it(`Should deauthenticate user in case of 401 return from server`, () => {
    const props = {
      service: {
        getTasks: (_) => ({
          fork: (rej, res) => {
            rej({ statusCode: 401 })
          }
        })
      },
      deauthenticate: sinon.spy()
    }

    const component = shallow(
      <TaskList {...props} />,
    );

    expect(props.deauthenticate.called).toBeTruthy();
  });
});
