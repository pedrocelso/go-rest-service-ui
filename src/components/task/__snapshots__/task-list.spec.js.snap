// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TaskList /> Should fetch the tasks from service and render 1`] = `
<div
  className="panel panel-default"
>
  <div
    className="list-group"
  >
    <div
      className="list-group-item d-flex row"
    >
      <div
        className="col-md-7"
      >
        <h5>
          Task Details
        </h5>
      </div>
      <div
        className="col-md-2"
      >
        <h5>
          Created at
        </h5>
      </div>
      <div
        className="col-md-2"
      >
        <h5>
          Updated at
        </h5>
      </div>
      <div
        className="col-md-1"
      >
        <h5>
          Inccidents
        </h5>
      </div>
    </div>
    <TaskItem
      key="10"
      task={
        Object {
          "description": "test@test.com",
          "id": 10,
          "name": "test",
        }
      }
    />
    <TaskItem
      key="11"
      task={
        Object {
          "description": "test1@test.com",
          "id": 11,
          "name": "test1",
        }
      }
    />
  </div>
</div>
`;

exports[`<TaskList /> Should not display tasks if service returns an error 1`] = `
<div
  className="panel panel-default"
/>
`;
